### DISCUSSION TITLE
How to define a dictionary property in a custom activity?

### DISCUSSION THREAD

**Discussion by VahidShir** [elsa-workflows/elsa-core#1250] [Q&A]
I already know how to [define simple property](https://elsa-workflows.github.io/elsa-core/docs/next/extensibility/extensibility-custom-activities#properties) types such as `string` in custom activities. It works fine and it is shown in designer and also exposed in HTTP endpoint JSON response as well, but I want a dictionary type, enabling us to fill it in activity designer.

Here's what I tried:
```
[Activity(
    DisplayName = "Custom Activity",
    Description = "It's a customized activity",
    Category = "Custom",
    Outcomes = new[] { OutcomeNames.Done }
)]
public class SampleActivity : Activity
{
    [ActivityProperty(Hint = "Variables associated to this activity")]
    public Dictionary<string, string> Variables { get; set; }

    protected override IActivityExecutionResult OnExecute(ActivityExecutionContext context)
    {
        Console.WriteLine("Variables: " + string.Join(", ", Variables.Keys));

        return Done();
    }
}
```
![image](https://user-images.githubusercontent.com/8729558/125185666-5b70d080-e23b-11eb-9ab8-8d025e5b4a80.png)

But this property isn't working as expected and I'm not able to insert anything into it.

- And also please tell me how to define a property in a custom activity without exposing it in designer, only exposing it in HTTP endpoints JSON results (setting it programmatically in activity).
Thanks.

**Comment by jdevillard** [comment]
Hello, 

have you take a look at the UIHint and SupportedSyntaxes  of the ActivityInput Decorator ? 

for example, 
```
        [ActivityInput(
            UIHint = ActivityInputUIHints.MultiLine,
            DefaultSyntax = SyntaxNames.Json,
            SupportedSyntaxes = new[] { SyntaxNames.JavaScript, SyntaxNames.Liquid, SyntaxNames.Json }
```

and try to write your json dictionnary in the Text Form (in JSON mode or Javascript mode)
```
{
"key1":"value1",
"key2":"value2":
}
```

**Comment by grigoris-adamopoulos** [comment]
Hello,

It seems that now Dictionary is available in ActivityInputUIHints:
for example:

```
[ActivityInput(            
    UIHint = ActivityInputUIHints.Dictionary,
    SupportedSyntaxes = new[] { SyntaxNames.JavaScript, SyntaxNames.Liquid }
)]
public Dictionary<string, string> Dimensions { get; set; }
```

In the UI of the activity everything seems fine: 

![image](https://user-images.githubusercontent.com/108006250/230335909-8bfe09b6-9358-4d95-acef-27497099be1c.png)

But during the execution of the activity, the variable `Dimensions` is null and is not resolved to the values specified.
Am I doing something wrong in the code?

**Reply by sfmskywalker** [reply]
Sounds like it might be a bug. I don't recall ever having used this, so I wouldn't be surprised if something was off. Please feel free to open an issue about this.

**Reply by vnmnm** [reply]
@grigoris-adamopoulos I was getting null too. Below decoration fixed the issue for me. I briefly checked the code and seems like the JSON Expression Evaluator can only deserialize this as expected. Other expression evaluators will fail to convert this as expected and hence returning null.

```
[ActivityInput(Hint = "Fields to update", UIHint = ActivityInputUIHints.Dictionary, DefaultSyntax = SyntaxNames.Json, SupportedSyntaxes = new[] { SyntaxNames.Json })]
public Dictionary<string, object> FieldsToUpdate { get; set; }
```
